File: rust/src/config/labels.rs
//! Label store: manages pubkey → labels mapping from multiple sources.
//!
//! Labels come from:
//! - Host LAN labels: config peers (e.g., "host.zigor.net/trusted")
//! - Remote LAN labels: zgrlan API (e.g., "company.zigor.net/employee")

use std::collections::HashMap;
use std::sync::RwLock;

/// Thread-safe store mapping pubkey hex strings to labels.
pub struct LabelStore {
    inner: RwLock<HashMap<String, Vec<String>>>,
}

impl Default for LabelStore {
    fn default() -> Self {
        Self::new()
    }
}

impl LabelStore {
    /// Create an empty LabelStore.
    pub fn new() -> Self {
        Self {
            inner: RwLock::new(HashMap::new()),
        }
    }

    /// Returns all labels for the given pubkey (hex-encoded, lowercase).
    pub fn labels(&self, pubkey_hex: &str) -> Vec<String> {
        let inner = self.inner.read().unwrap();
        inner.get(pubkey_hex).cloned().unwrap_or_default()
    }

    /// Add labels to the given pubkey. Duplicates are ignored.
    pub fn add_labels(&self, pubkey_hex: &str, labels: &[String]) {
        if labels.is_empty() {
            return;
        }

        let mut inner = self.inner.write().unwrap();
        let existing = inner.entry(pubkey_hex.to_string()).or_default();
        for label in labels {
            if !existing.contains(label) {
                existing.push(label.clone());
            }
        }
    }

    /// Replace all labels for the given pubkey.
    pub fn set_labels(&self, pubkey_hex: &str, labels: &[String]) {
        let mut inner = self.inner.write().unwrap();
        if labels.is_empty() {
            inner.remove(pubkey_hex);
        } else {
            inner.insert(pubkey_hex.to_string(), labels.to_vec());
        }
    }

    /// Remove all labels for the given pubkey that belong to the
    /// specified LAN domain prefix. For example, `remove_labels(pk, "company.zigor.net")`
    /// removes all labels starting with "company.zigor.net/".
    pub fn remove_labels(&self, pubkey_hex: &str, lan_domain: &str) {
        let prefix = format!("{}/", lan_domain);

        let mut inner = self.inner.write().unwrap();
        if let Some(existing) = inner.get_mut(pubkey_hex) {
            existing.retain(|l| !l.starts_with(&prefix));
            if existing.is_empty() {
                inner.remove(pubkey_hex);
            }
        }
    }

    /// Remove all labels for the given pubkey.
    pub fn remove_peer(&self, pubkey_hex: &str) {
        let mut inner = self.inner.write().unwrap();
        inner.remove(pubkey_hex);
    }

    /// Populate the store with labels from config peers.
    /// Preserves existing labels from other sources (e.g., remote LANs).
    pub fn load_from_config(&self, peers: &std::collections::HashMap<String, super::PeerConfig>) {
        let mut inner = self.inner.write().unwrap();

        for (domain, peer) in peers {
            let pubkey_hex = match pubkey_hex_from_domain(domain) {
                Some(pk) => pk,
                None => continue,
            };
            if peer.labels.is_empty() {
                continue;
            }

            // Remove existing host lan labels, keep remote lan labels
            let existing = inner.entry(pubkey_hex.clone()).or_default();
            existing.retain(|l| !l.starts_with("host.zigor.net/"));

            // Add config labels
            for label in &peer.labels {
                if !existing.contains(label) {
                    existing.push(label.clone());
                }
            }
        }
    }

    /// Returns the number of pubkeys with labels.
    pub fn count(&self) -> usize {
        let inner = self.inner.read().unwrap();
        inner.len()
    }
}

/// Check if any of the peer's labels match a label pattern.
///
/// Supports:
/// - Exact match: "host.zigor.net/trusted"
/// - Wildcard match: "company.zigor.net/*" (matches any label under that domain)
pub fn match_label(peer_labels: &[String], pattern: &str) -> bool {
    if let Some(prefix) = pattern.strip_suffix("/*") {
        let prefix_with_slash = format!("{}/", prefix);
        peer_labels.iter().any(|l| l.starts_with(&prefix_with_slash))
    } else {
        peer_labels.iter().any(|l| l == pattern)
    }
}

/// Check if any of the peer's labels match any of the patterns.
/// Returns true if at least one pattern matches.
pub fn match_labels(peer_labels: &[String], patterns: &[String]) -> bool {
    patterns.iter().any(|p| match_label(peer_labels, p))
}

/// Extract the hex pubkey from a peer domain.
/// Format: "{hex}.zigor.net" -> lowercase hex string.
fn pubkey_hex_from_domain(domain: &str) -> Option<String> {
    let prefix = domain.strip_suffix(".zigor.net")?;
    if prefix.is_empty() || prefix.len() > 64 {
        return None;
    }
    let lower = prefix.to_lowercase();
    if lower.chars().all(|c| c.is_ascii_hexdigit()) {
        Some(lower)
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_operations() {
        let store = LabelStore::new();
        let pk = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";

        // Initially empty
        assert!(store.labels(pk).is_empty());
        assert_eq!(store.count(), 0);

        // Add labels
        store.add_labels(pk, &[
            "host.zigor.net/trusted".into(),
            "host.zigor.net/friend".into(),
        ]);
        let labels = store.labels(pk);
        assert_eq!(labels.len(), 2);
        assert_eq!(labels[0], "host.zigor.net/trusted");
        assert_eq!(labels[1], "host.zigor.net/friend");
        assert_eq!(store.count(), 1);
    }

    #[test]
    fn test_add_duplicates() {
        let store = LabelStore::new();
        let pk = "0000000000000000000000000000000000000000000000000000000000000001";

        store.add_labels(pk, &["host.zigor.net/trusted".into()]);
        store.add_labels(pk, &[
            "host.zigor.net/trusted".into(),
            "host.zigor.net/friend".into(),
        ]);

        let labels = store.labels(pk);
        assert_eq!(labels.len(), 2);
    }

    #[test]
    fn test_set_labels() {
        let store = LabelStore::new();
        let pk = "0000000000000000000000000000000000000000000000000000000000000001";

        store.add_labels(pk, &["host.zigor.net/old".into()]);
        store.set_labels(pk, &["host.zigor.net/new".into()]);

        let labels = store.labels(pk);
        assert_eq!(labels, vec!["host.zigor.net/new"]);

        // Empty set removes peer
        store.set_labels(pk, &[]);
        assert!(store.labels(pk).is_empty());
        assert_eq!(store.count(), 0);
    }

    #[test]
    fn test_remove_by_lan_domain() {
        let store = LabelStore::new();
        let pk = "0000000000000000000000000000000000000000000000000000000000000001";

        store.add_labels(pk, &[
            "host.zigor.net/trusted".into(),
            "company.zigor.net/employee".into(),
            "company.zigor.net/dev-team".into(),
        ]);

        store.remove_labels(pk, "company.zigor.net");

        let labels = store.labels(pk);
        assert_eq!(labels, vec!["host.zigor.net/trusted"]);
    }

    #[test]
    fn test_remove_peer() {
        let store = LabelStore::new();
        let pk = "0000000000000000000000000000000000000000000000000000000000000001";

        store.add_labels(pk, &["host.zigor.net/trusted".into()]);
        store.remove_peer(pk);

        assert!(store.labels(pk).is_empty());
    }

    #[test]
    fn test_load_from_config() {
        let store = LabelStore::new();
        let pk = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567aa";
        let domain = format!("{}.zigor.net", pk);

        // Pre-populate with remote lan labels
        store.add_labels(pk, &["company.zigor.net/employee".into()]);

        let mut peers = HashMap::new();
        peers.insert(domain, super::super::PeerConfig {
            alias: "peer_us".into(),
            direct: vec!["1.2.3.4:51820".into()],
            relay: vec![],
            labels: vec!["host.zigor.net/trusted".into(), "host.zigor.net/exit-node".into()],
        });

        store.load_from_config(&peers);

        let labels = store.labels(pk);
        assert_eq!(labels.len(), 3);
        assert!(labels.contains(&"company.zigor.net/employee".to_string()));
        assert!(labels.contains(&"host.zigor.net/trusted".to_string()));
        assert!(labels.contains(&"host.zigor.net/exit-node".to_string()));
    }

    #[test]
    fn test_match_label_exact() {
        let labels: Vec<String> = vec![
            "host.zigor.net/trusted".into(),
            "company.zigor.net/employee".into(),
        ];

        assert!(match_label(&labels, "host.zigor.net/trusted"));
        assert!(!match_label(&labels, "host.zigor.net/friend"));
    }

    #[test]
    fn test_match_label_wildcard() {
        let labels: Vec<String> = vec![
            "company.zigor.net/employee".into(),
            "company.zigor.net/dev-team".into(),
        ];

        assert!(match_label(&labels, "company.zigor.net/*"));
        assert!(!match_label(&labels, "other.zigor.net/*"));
    }

    #[test]
    fn test_match_labels() {
        let labels: Vec<String> = vec!["host.zigor.net/trusted".into()];

        assert!(match_labels(&labels, &[
            "company.zigor.net/*".into(),
            "host.zigor.net/trusted".into(),
        ]));

        assert!(!match_labels(&labels, &[
            "company.zigor.net/*".into(),
            "other.zigor.net/admin".into(),
        ]));
    }

    #[test]
    fn test_match_labels_empty() {
        assert!(!match_labels(&[], &["host.zigor.net/trusted".into()]));
        assert!(!match_labels(&["host.zigor.net/trusted".into()], &[]));
    }

    #[test]
    fn test_pubkey_hex_from_domain() {
        assert_eq!(
            pubkey_hex_from_domain("abcdef01.zigor.net"),
            Some("abcdef01".into())
        );
        assert_eq!(
            pubkey_hex_from_domain("ABCDEF01.zigor.net"),
            Some("abcdef01".into())
        );
        assert_eq!(pubkey_hex_from_domain("not-hex.zigor.net"), None);
        assert_eq!(pubkey_hex_from_domain("example.com"), None);
        assert_eq!(pubkey_hex_from_domain(".zigor.net"), None);
    }
}

File: zig/src/config/labels.zig
//! Label store: manages pubkey → labels mapping from multiple sources.
//!
//! Labels come from:
//! - Host LAN labels: config peers (e.g., "host.zigor.net/trusted")
//! - Remote LAN labels: zgrlan API (e.g., "company.zigor.net/employee")

const std = @import("std");
const mem = std.mem;
const Allocator = std.mem.Allocator;

/// Thread-safe store mapping pubkey hex strings to labels.
pub const LabelStore = struct {
    allocator: Allocator,
    /// pubkey hex -> list of labels (all allocator-owned strings)
    labels: std.StringHashMapUnmanaged(std.ArrayListUnmanaged([]const u8)),

    pub fn init(allocator: Allocator) LabelStore {
        return .{
            .allocator = allocator,
            .labels = .{},
        };
    }

    pub fn deinit(self: *LabelStore) void {
        var it = self.labels.iterator();
        while (it.next()) |entry| {
            for (entry.value_ptr.items) |label| {
                self.allocator.free(label);
            }
            entry.value_ptr.deinit(self.allocator);
            self.allocator.free(entry.key_ptr.*);
        }
        self.labels.deinit(self.allocator);
    }

    /// Returns all labels for the given pubkey hex.
    pub fn getLabels(self: *const LabelStore, pubkey_hex: []const u8) []const []const u8 {
        if (self.labels.get(pubkey_hex)) |list| {
            return list.items;
        }
        return &.{};
    }

    /// Add labels to the given pubkey. Duplicates are ignored.
    pub fn addLabels(self: *LabelStore, pubkey_hex: []const u8, new_labels: []const []const u8) !void {
        if (new_labels.len == 0) return;

        const gop = try self.labels.getOrPut(self.allocator, pubkey_hex);
        if (!gop.found_existing) {
            gop.key_ptr.* = try self.allocator.dupe(u8, pubkey_hex);
            gop.value_ptr.* = .{};
        }
        var list = gop.value_ptr;

        for (new_labels) |label| {
            var found = false;
            for (list.items) |existing| {
                if (mem.eql(u8, existing, label)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                const owned = try self.allocator.dupe(u8, label);
                try list.append(self.allocator, owned);
            }
        }
    }

    /// Replace all labels for the given pubkey.
    pub fn setLabels(self: *LabelStore, pubkey_hex: []const u8, new_labels: []const []const u8) !void {
        if (self.labels.getPtr(pubkey_hex)) |list| {
            for (list.items) |label| {
                self.allocator.free(label);
            }
            list.clearRetainingCapacity();

            if (new_labels.len == 0) {
                list.deinit(self.allocator);
                const key = self.labels.getKey(pubkey_hex).?;
                _ = self.labels.remove(pubkey_hex);
                self.allocator.free(key);
                return;
            }

            for (new_labels) |label| {
                const owned = try self.allocator.dupe(u8, label);
                try list.append(self.allocator, owned);
            }
        } else if (new_labels.len > 0) {
            try self.addLabels(pubkey_hex, new_labels);
        }
    }

    /// Remove all labels matching a LAN domain prefix (e.g., "company.zigor.net").
    pub fn removeLabels(self: *LabelStore, pubkey_hex: []const u8, lan_domain: []const u8) !void {
        const list = self.labels.getPtr(pubkey_hex) orelse return;

        // Build prefix: "company.zigor.net/"
        const prefix = try std.fmt.allocPrint(self.allocator, "{s}/", .{lan_domain});
        defer self.allocator.free(prefix);

        var i: usize = 0;
        while (i < list.items.len) {
            if (mem.startsWith(u8, list.items[i], prefix)) {
                self.allocator.free(list.items[i]);
                _ = list.swapRemove(i);
            } else {
                i += 1;
            }
        }

        if (list.items.len == 0) {
            list.deinit(self.allocator);
            const key = self.labels.getKey(pubkey_hex).?;
            _ = self.labels.remove(pubkey_hex);
            self.allocator.free(key);
        }
    }

    /// Remove all labels for the given pubkey.
    pub fn removePeer(self: *LabelStore, pubkey_hex: []const u8) void {
        if (self.labels.getPtr(pubkey_hex)) |list| {
            for (list.items) |label| {
                self.allocator.free(label);
            }
            list.deinit(self.allocator);
            const key = self.labels.getKey(pubkey_hex).?;
            _ = self.labels.remove(pubkey_hex);
            self.allocator.free(key);
        }
    }

    /// Returns the number of pubkeys with labels.
    pub fn count(self: *const LabelStore) usize {
        return self.labels.count();
    }
};

/// Check if any of the peer's labels match a label pattern.
/// Supports exact match and wildcard ("company.zigor.net/*").
pub fn matchLabel(peer_labels: []const []const u8, pattern: []const u8) bool {
    // Check for wildcard: ends with "/*"
    if (pattern.len >= 2 and mem.endsWith(u8, pattern, "/*")) {
        const domain_prefix = pattern[0 .. pattern.len - 1]; // "company.zigor.net/"
        for (peer_labels) |label| {
            if (mem.startsWith(u8, label, domain_prefix)) {
                return true;
            }
        }
        return false;
    }

    // Exact match
    for (peer_labels) |label| {
        if (mem.eql(u8, label, pattern)) {
            return true;
        }
    }
    return false;
}

/// Check if any of the peer's labels match any of the patterns.
pub fn matchLabels(peer_labels: []const []const u8, patterns: []const []const u8) bool {
    for (patterns) |pattern| {
        if (matchLabel(peer_labels, pattern)) {
            return true;
        }
    }
    return false;
}

// =========================================================================
// Tests
// =========================================================================

test "LabelStore basic operations" {
    var store = LabelStore.init(std.testing.allocator);
    defer store.deinit();

    const pk = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";

    // Initially empty
    try std.testing.expectEqual(@as(usize, 0), store.getLabels(pk).len);
    try std.testing.expectEqual(@as(usize, 0), store.count());

    // Add labels
    try store.addLabels(pk, &.{ "host.zigor.net/trusted", "host.zigor.net/friend" });
    const labels = store.getLabels(pk);
    try std.testing.expectEqual(@as(usize, 2), labels.len);
    try std.testing.expectEqualStrings("host.zigor.net/trusted", labels[0]);
    try std.testing.expectEqualStrings("host.zigor.net/friend", labels[1]);
    try std.testing.expectEqual(@as(usize, 1), store.count());
}

test "LabelStore add duplicates" {
    var store = LabelStore.init(std.testing.allocator);
    defer store.deinit();

    const pk = "0000000000000000000000000000000000000000000000000000000000000001";
    try store.addLabels(pk, &.{"host.zigor.net/trusted"});
    try store.addLabels(pk, &.{ "host.zigor.net/trusted", "host.zigor.net/friend" });

    try std.testing.expectEqual(@as(usize, 2), store.getLabels(pk).len);
}

test "LabelStore remove peer" {
    var store = LabelStore.init(std.testing.allocator);
    defer store.deinit();

    const pk = "0000000000000000000000000000000000000000000000000000000000000001";
    try store.addLabels(pk, &.{"host.zigor.net/trusted"});
    store.removePeer(pk);

    try std.testing.expectEqual(@as(usize, 0), store.getLabels(pk).len);
    try std.testing.expectEqual(@as(usize, 0), store.count());
}

test "matchLabel exact" {
    const labels = [_][]const u8{ "host.zigor.net/trusted", "company.zigor.net/employee" };
    try std.testing.expect(matchLabel(&labels, "host.zigor.net/trusted"));
    try std.testing.expect(!matchLabel(&labels, "host.zigor.net/friend"));
}

test "matchLabel wildcard" {
    const labels = [_][]const u8{ "company.zigor.net/employee", "company.zigor.net/dev-team" };
    try std.testing.expect(matchLabel(&labels, "company.zigor.net/*"));
    try std.testing.expect(!matchLabel(&labels, "other.zigor.net/*"));
}

test "matchLabels" {
    const labels = [_][]const u8{"host.zigor.net/trusted"};
    const patterns = [_][]const u8{ "company.zigor.net/*", "host.zigor.net/trusted" };
    try std.testing.expect(matchLabels(&labels, &patterns));

    const patterns2 = [_][]const u8{ "company.zigor.net/*", "other.zigor.net/admin" };
    try std.testing.expect(!matchLabels(&labels, &patterns2));
}

test "matchLabels empty" {
    const empty = [_][]const u8{};
    const patterns = [_][]const u8{"host.zigor.net/trusted"};
    try std.testing.expect(!matchLabels(&empty, &patterns));
    const labels = [_][]const u8{"host.zigor.net/trusted"};
    try std.testing.expect(!matchLabels(&labels, &empty));
}
